### PR TITLE
Add E2E test case cover duplicatepods strategy

### PR DESCRIPTION
**Pull request by JaneLiuL**
Add duplicatepods strategy into E2E test case :)

### CODE REVIEW

**Comment by a7i** [comment]
What are your thoughts on moving this to its own file? `e2e_toomanyrestarts_test.go`?

I followed [this](https://github.com/kubernetes-sigs/descheduler/blob/master/test/e2e/e2e_topologyspreadconstraint_test.go) pattern for TopologySpreadConstraint strategy as a single file was making it hard to read/follow.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:2]: 2021

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:29]: Can you group this import with the `k8s.io` prefixed ones that are above?

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:130]: `RcByNameContainer` creates a pod template without the pod.spec set. So the test must fail in any case. Also, I don't think `clientSet.CoreV1().Pods(pod.Namespace).Create` will create a pod with the pod.Status as provided (maybe I am wrong?). One needs to use `Pods(...).Status(...).Update subresource instead. 

Can you move `t.Log("Creating pods with different restartCount")` part into `for _, tc := range tests` and have the `RestartCount` as part of the test struct? Then creating the pods with the `RestartCount` in every test iteration right before `RcByNameContainer` is invoked?

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:157]: `workerNodes[0]` is too specific. In different cluster with many worker nodes the pods may be scheduled to different worker nodes than the first one. Can you update the code to list all relevant pods from all the worker nodes instead?

**Review by ingvagabund — CHANGES_REQUESTED** [review]
Thanks for the test!!! I dropped some comments.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_toomanyrestarts_test.go:2]: Haha, I will fix, but actually I am really confuse that other files use 2017.. Could I fix all others files to 2021 or only let it go?

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_toomanyrestarts_test.go:157]: Thanks, I check the E2E Test infra, it only create a Kind cluster, no other nodes join, that's why I use workerNode[0], and I also reference other test code, they use the  same word.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_toomanyrestarts_test.go:130]: Yes, I will fix, much thanks :)

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [reply test/e2e/e2e_toomanyrestarts_test.go:2]: Yeah, the other files are a bit pre-historic. When creating new files we tend to use the current year. Actually, there's no test checking if all files have the right header so there are even some files without it. It's sufficient to just change the year in this file.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [reply test/e2e/e2e_toomanyrestarts_test.go:157]: If you find a way to write the code independent of the number of worker nodes the better. Otherwise, it's not required. Just better to have.

**Comment by JaneLiuL** [comment]
I am really confuse for the test, I am so sorry that it fails, but my testcase already pass, it fails other test case, I don't change other test case...
```
=== RUN   TestRemoveDuplicates
    e2e_duplicatepods_test.go:51: Creating testing namespace
    e2e_duplicatepods_test.go:66: Creating duplicates pods
    e2e_duplicatepods_test.go:113: Creating deployment duplicate-pod in e2e-testremoveduplicates namespace
    e2e_duplicatepods_test.go:186: Waiting for 4 pods to be created, got 0 instead
    e2e_duplicatepods_test.go:150: Running DeschedulerStrategy strategy
    e2e_duplicatepods_test.go:173: Number of pods on node kind-worker changed from 4 to 4
--- **PASS: TestRemoveDuplicates (10.12s**

=== RUN   TestTopologySpreadConstraint/test-rc-topology-spread-hard-constraint
    e2e_topologyspreadconstraint_test.go:61: Creating RC test-rc-topology-spread-hard-constraint with 4 replicas
    e2e_test.go:980: Waiting for 4 pods to be created, got 0 instead
    e2e_test.go:980: Waiting for 2 pods to be created, got 0 instead
    e2e_test.go:985: Pod test-rc-topology-spread-hard-constraint-violator-b8jbl not running yet, is Pending instead
    e2e_test.go:985: Pod test-rc-topology-spread-hard-constraint-violator-b8jbl not running yet, is Pending instead
...
--- **FAIL: TestTopologySpreadConstraint** (94.29s)
    --- FAIL: TestTopologySpreadConstraint/test-rc-topology-spread-soft-constraint (47.13s)
    --- FAIL: TestTopologySpreadConstraint/test-rc-topology-spread-hard-constraint (47.13s)
```

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_toomanyrestarts_test.go]: I try to use `if _, err := clientSet.CoreV1().Pods(pod.Namespace).UpdateStatus(ctx, pod, metav1.UpdateOptions{}); err != nil {` to update the restartCount, it success, but very soon, about 2 seconds, the kubelet will rollback the restartCount back to 0.
So I change to only test duplicatePods.

Comment (ingvagabund) [reply test/e2e/e2e_toomanyrestarts_test.go]: The kubelet will not let himself to be fooled quite quickly. You might create pods that will constantly crash (e.g. running `sleep 1s && exit 1`).

File: test/e2e/e2e_duplicatepods_test.go
Comment (ingvagabund) [inline test/e2e/e2e_duplicatepods_test.go:92]: Do you plan to add more tests before the PR merges?

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_toomanyrestarts_test.go]: very good idea~ Before your comment, I just use `kubectl exec -it %s -n %s -- /bin/sh -c "kill 1"` for kill pod :( 
I will fix soon~ Thanks a lot~

File: test/e2e/e2e_duplicatepods_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_duplicatepods_test.go:92]: I still need time for enhance the toomanyRestart test case :)
Which one you prefer? One PR or another PR to cover toomanyRestart test case?

Comment (ingvagabund) [reply test/e2e/e2e_duplicatepods_test.go:92]: There's still time to do both :) It will take a time before we release the descheduler again.

Comment (ingvagabund) [reply test/e2e/e2e_duplicatepods_test.go:92]: Each new test in its own PR is preferable

Comment (JaneLiuL) [reply test/e2e/e2e_duplicatepods_test.go:92]: Hi , I finish the toomanyrestart test case and all success now :)

**Comment by JaneLiuL** [comment]
@ingvagabund I already fix all the comments and also including the toomangrestart test strategy.
Would you help to review ? :) Much thanks

**Comment by ingvagabund** [comment]
On my todo list :)

File: test/e2e/e2e_duplicatepods_test.go
Comment (ingvagabund) [inline test/e2e/e2e_duplicatepods_test.go:176]: `s/created/running`

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:169]: Could you use `wait.PollImmediate` instead of for+select? `wait.PollImmediate` will give you the same functionality, just a bit cleaner.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:160]: Let's use "!=" to report even the case when more pods than expected are evicted.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:161]: Can you also mention `tc.expectedEvictedPodCount` in the error message?

**Review by ingvagabund — CHANGES_REQUESTED** [review]
Looks good in overall :) Just a couple of nits.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_toomanyrestarts_test.go:169]: Hi. I agree with you. At first I use `wait.PollImmediate`, but soon I give up.
That's for I need to execute List pods many times and util all pod restart count more than 4.
But I I use `wait.PollImmediate`, I have to return when I execute the first time to List Pods.
So I can use `for` for I need to execute list pods many times.

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_toomanyrestarts_test.go]: fix done~

File: test/e2e/e2e_duplicatepods_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_duplicatepods_test.go]: Yes. fix done~

File: test/e2e/e2e_toomanyrestarts_test.go
Comment (ingvagabund) [reply test/e2e/e2e_toomanyrestarts_test.go:161]: Sounds reasonable, thanks for the clarification.

**Comment by JaneLiuL** [comment]
So sorry to at you again. It seems still need /approve label could be merge :) @ingvagabund

**Comment by ingvagabund** [comment]
I asked @damemi for the final approval :) Just in case I missed something

File: test/e2e/e2e_duplicatepods_test.go
Comment (a7i) [inline test/e2e/e2e_duplicatepods_test.go:120]: You probably want to run the tests separately by utilizing:

```
	for _, tc := range tests {
		t.Run(tc.description, , func(t *testing.T) {
		
		}
	}	
```

Otherwise, you won't know which test case failed unless you read every log line. same in too many restarts.

File: test/e2e/e2e_duplicatepods_test.go
Comment (a7i) [inline test/e2e/e2e_duplicatepods_test.go:133]: I would think this should be uncommented to prevent cascading failures if one test fails.

File: test/e2e/e2e_duplicatepods_test.go
Comment (a7i) [inline test/e2e/e2e_duplicatepods_test.go:172]: better to do this up top and use defer

Comment (JaneLiuL) [reply test/e2e/e2e_duplicatepods_test.go]: Fix done. Would you please kindly review?

File: test/e2e/e2e_duplicatepods_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_duplicatepods_test.go]: Fix done. Would you please kindly review?

File: test/e2e/e2e_duplicatepods_test.go
Comment (JaneLiuL) [reply test/e2e/e2e_duplicatepods_test.go]: Fix done. Would you please kindly review?

**Comment by a7i** [comment]
> @ingvagabund @a7i So sorry to at you again, would you help to review and approve if no other comments? :)

LGTM but I'm not a codeowner

**Pull request merged**
